query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -35,14 +38,83 @@
   // while the selectedColumns
   // is a set of column names.
   private final LinkedHashSet<Expression> allSelections;
+  private final Map<String,String> timeSeriesColumnMap;
+  private final Map<String,List<Long>> timeFilterMap;
 
   public ExecutionContext(String tenantId, QueryRequest request) {
     this.tenantId = tenantId;
     this.selectedColumns = new LinkedHashSet<>();
     this.allSelections = new LinkedHashSet<>();
+    this.timeSeriesColumnMap = new HashMap<>();
+    this.timeFilterMap = new HashMap<String,List<Long>>();
+    analyzeForAvgRate(request);
     analyze(request);
   }
 
+  private void analyzeForAvgRate(QueryRequest request){
+    setTimeSeriesColumnMap(request);
+    setTimeFilterMap(request);
+  }
+
+  private void setTimeSeriesColumnMap(QueryRequest request){
+
+    if(request.getGroupByCount() > 0){
+      for(Expression expression : request.getGroupByList()){
+        if(expression.getValueCase() == ValueCase.FUNCTION && expression.getFunction().getFunctionName().equals("dateTimeConvert")) {
+
+          String column = null;
+          String period = null;
+
+          //assuming only one column and one literal (with period in seconds) is present in one dateTimeConvert groupBy
+          for(Expression childExpression : expression.getFunction().getArgumentsList()) {
+            if(childExpression.getValueCase() == ValueCase.COLUMNIDENTIFIER) {
+              column = childExpression.getColumnIdentifier().getColumnName().split("[.]")[0];
+            }
+            if(childExpression.getValueCase() == ValueCase.LITERAL){
+              if(childExpression.getLiteral().getValue().getString().split("[:]")[1].equals("SECONDS")){
+                period = childExpression.getLiteral().getValue().getString().split("[:]")[0];
+              }
+            }
+          }
+
+          if(column != null && period != null){
+            timeSeriesColumnMap.put(column,period);
+          }
+        }
+      }
+    }
+  }
+
+  private void setTimeFilterMap(QueryRequest request) {
+
+    if(request.getFilter().getChildFilterCount() > 0){
+      for(Filter filter : request.getFilter().getChildFilterList()){
+
+        String[] columnName = filter.getLhs().getColumnIdentifier().getColumnName().split("[.]");
+        Operator operator = filter.getOperator();
+        Long val = filter.getRhs().getLiteral().getValue().getLong();
+
+        //filtering by colName.startTime and colName.endTime for now. Can be changed accordingly
+        if(columnName.length > 1 && (columnName[1].equals("startTime") || columnName[1].equals("endTime"))){
```

Comment:
this is a big assumption, these column names don't hold for every view and should not be assumed. Not sure if this can be looked up - pretty sure it's specified somewhere - or if we have to drive it via config.

Reply:
Can be extracted from a config file. Will be done as a follow up PR